### ISSUE TITLE
Rotating list losing key order

### ISSUE THREAD

**Issue by echenley** [chenglou/react-motion#264]
I have a `RecentNotes` component that lists the 5 most recent notes. When I add a note, the last one is removed, and when I remove a note, another note should appear at the end of the list. I am using Redux and Redux Undo in this example. I noticed that when I undo a removal, the new element that was added will jump to the location of the removed element, instead of remaining in its correct order:

![react-motion](https://cloud.githubusercontent.com/assets/6344219/12526993/e626b10a-c139-11e5-927e-e05ebc8c8b83.gif)

In this example the `notes` prop changes from :
```js
[{ obj: 1 }, { obj: 2 }, ... , { obj: 5 }]
```
to:
```js
[{ obj: 2 }, { obj: 3 }, ... , { obj: 6 }]
```
And back to:

```js
[{ obj: 1 }, { obj: 2 }, ... , { obj: 5 }]
```

During this last step `{ obj: 6 }` jumps to the beginning of the list.

How am I able to manage the order of elements in this type of scenario? This is my full implementation:

```js
class NoteList extends Component {
    constructor(props) {
        super(props);
    }

    willEnter() {
        return {
            height: 0,
            opacity: 0
        };
    }

    willLeave() {
        return {
            height: spring(0),
            opacity: spring(0)
        };
    }

    render() {
        const {
            notes,
            accountId,
            actions,
            routeActions,
            ...props
        } = this.props;

        return (
            <div { ...props }>
                <TransitionMotion
                    styles={ notes.map(note => ({
                        key: note.id,
                        style: { height: spring(styles.note.height), opacity: spring(1) },
                        data: note
                    })) }
                    willEnter={ this.willEnter.bind(this) }
                    willLeave={ this.willLeave.bind(this) }
                >
                    { interpolatedStyles =>
                        <div>
                            { interpolatedStyles.map(({ key, data, style }) =>
                                <div key={ key } style={ { ...style, ...styles.noteWrap } }>
                                    <NoteListItem note={ data } />
                                </div>
                            ) }
                        </div>
                    }
                </TransitionMotion>
            </div>
        );
    }
}
```

**Comment by echenley** [comment]
[React Motion UI Pack](https://github.com/souporserious/react-motion-ui-pack) is able to do this sort of thing, so it must have to do with my implementation. Closing issue.

**Comment by echenley** [comment]
After further investigation, it looks like the issue still persists even using react-motion-ui-pack. I updated to the latest react-motion release (v0.4.1) and modified the code in my first comment to use the new API. Here is another gif of the issue:

![react-motion2](https://cloud.githubusercontent.com/assets/6344219/12595407/b53cf554-c440-11e5-8db1-ef031078bcc1.gif)

Strangely, if I hit undo quick enough (before the spring is completed), the last note keeps its position and everything works as intended. If the spring is allowed to complete, the last note will jump to the wrong position.

It seems to have to do when one item enters and another leaves simultaneously.

**Comment by chenglou** [comment]
Humm, looks like a key issue? What's your key?

**Comment by echenley** [comment]
Keys are unique ids which look like:

```js
const noteIds = notes.map(note => note.id)
// => ["569e86aa5dedf4004c8da568", "569e93045dedf4004c8da56a", "569fb4065dedf4088415caf8", "569ffe935dedf4088415caf9", "569ffe9f5dedf4088415cafa", "56a13afb5dedf40a4866c605", "56a1560e5dedf40a60a963d9"]
```

I am going to try to make a minimal repro of the issue. Stay tuned.

**Comment by echenley** [comment]
Here is a minimal repro: http://codepen.io/anon/pen/EPLyLW?editors=0010.

**Comment by chenglou** [comment]
Wow, you've discovered a rare (but theoretically important!) bug in the core keys merging algorithm. That's completely my fault. Sorry for taking so long to reply to this.

**Comment by echenley** [comment]
:+1: Happy to help! Thanks for being so responsive!

**Comment by chenglou** [comment]
No problem! Pushed as a fix in v0.4.2 just now. Small notice: I saw that in your real app you were using string key. Keep using these. I've reverted from letting anything be key, just like before.
